### PR TITLE
Introduce kvstoremesh, a clustermesh-apiserver companion component

### PR DESCRIPTION
**Pull request by giorio94**
Please review commit by commit. I'm reporting the description of the main commit here for your convenience:

```
In the current clustermesh implementation, each agent connects to the
kvstore in all remote clusters, pulling information about nodes,
services, identities and IPs. The number of etcd watchers can grow
significantly large (~ 5 * total nodes in the clustermesh), as well
as the number of events to be relayed by etcd to the agents (especially
if the churn rate is high).

kvstoremesh is a new component responsible for synchronizing the
information from the remote kvstores to the local one. In other words,
the local kvstore caches the state of the entire clustermesh, so that
all agents only need to connect to the local kvstore, and not the remote
ones. Overall, this provides better isolation (because the only component
connecting to remote kvstores is now the kvstoremesh operator) and
reduces the number of agents connecting to a single kvstore.

While the number of events to be processed by a single kvstore is
similar to the vanilla clustermesh case, with kvstoremesh the load
is less spiky (because a specific event is propagated to a much lower
number of agents) and it is also easier to rate limit (configuring the
maximum number of keys that can be written by kvstoremesh per second).
This allows to trade latency for reliability in case of high churn in
the clustermesh. Finally, given that the number of watchers is
proportional to the number of nodes in the local cluster only (rather
than the nodes in the entire clustermesh), the dimensioning of the
kvstore is also simplified.

The kvstoremesh concept has been initially proposed by the Cloud-Network
team @Trip.com, led by Arthur Chiao:
http://arthurchiao.art/blog/trip-first-step-towards-cloud-native-security/
```

The Dockerfile and CI modifications required to build the new image, as well as the appropriate extensions to the helm chart, will be introduced in a subsequent PR to limit the size of the current one.

```release-note
Introduce kvstoremesh, a clustermesh-apiserver companion component allowing to cache remote cluster information in the local kvstore for increased scalability and separation.
```

### CODE REVIEW

File: kvstoremesh/metrics/metrics.go
Comment (tommyp1ckles) [inline kvstoremesh/metrics/metrics.go:108]: The Agent server enables OpenMetrics in HandlerOpts, Im not that familiar with the option but might be something to consider to make sure metrics are consistent among the different components.

File: kvstoremesh/metrics/metrics.go
Comment (tommyp1ckles) [inline kvstoremesh/metrics/metrics.go]: In the Agent we recently enabled sched latency metrics: https://github.com/cilium/cilium/blob/main/pkg/metrics/registry.go#L116 by default.

The idea was that when people run Cilium in severely CPU constrained environments that we sometimes begin to see strange behaviour as goroutines get large scheduling latency.

Might be worth something to have here too, keep things consistent and we can easily build alerting around that.

File: kvstoremesh/metrics/metrics.go
Comment (tommyp1ckles) [inline kvstoremesh/metrics/metrics.go:80]: Should these be namespaced (i.e. `cilium_kvstoremesh`)?

File: kvstoremesh/metrics/metrics.go
Comment (giorio94) [reply kvstoremesh/metrics/metrics.go:108]: I could find `EnableOpenMetrics` being used as an option only for the [Hubble metrics](https://github.com/cilium/cilium/blob/7ec43fe8c6bf53bba6476132200aa1623806713b/pkg/hubble/metrics/metrics.go#L78-L79), but not for the [agent metrics](https://github.com/cilium/cilium/blob/7ec43fe8c6bf53bba6476132200aa1623806713b/pkg/metrics/registry.go#LL79C17-L79C17). Am I missing anything?

File: kvstoremesh/metrics/metrics.go
Comment (giorio94) [reply kvstoremesh/metrics/metrics.go:80]: It seems that there is currently a bit of inconsistency in the code base, since process-related metrics are currently namespaced in the agents and cilium-operator, and non-namespaced in hubble relay and in the clustermesh-apiserver.  Go related metrics instead are always non-namespaced. Do you feel it would be better change the process ones to be namespaced in clustermesh-apiserver/kvstoremesh?

File: kvstoremesh/metrics/metrics.go
Comment (giorio94) [reply kvstoremesh/metrics/metrics.go]: Added, thanks for the heads up.

File: kvstoremesh/metrics/metrics.go
Comment (tommyp1ckles) [reply kvstoremesh/metrics/metrics.go:80]: Good point, the agent/operator metrics declarations also all specify their namespace in-line as well. 
It looks like hubble metrics are just name spaced individually with the "hubble" namespace.

So from what I can tell, with reuising the existing pkg/metrics declarations this will result in using the "cilium_" namespace?

Given that we're reusing the same metric name I think it would make sense for this component to have its own namespace prefix to distinguish between agent/kvstoremesh controller metrics.

File: kvstoremesh/metrics/metrics.go
Comment (tommyp1ckles) [inline kvstoremesh/metrics/metrics.go:91]: Small optional point, it's always nice to add a "version" metrics such as `cilium_version` for a couple reasons:

* Discover ability: low cardinality, generally always enabled metric that can be used to detect running deployments/labels etc.
* Monitoring component versions via Prometheus.

File: kvstoremesh/metrics/metrics.go
Comment (tommyp1ckles) [reply kvstoremesh/metrics/metrics.go:108]: Yeah not sure if this is really necessary.

File: Documentation/observability/metrics.rst
Comment (zacharysarah) [inline Documentation/observability/metrics.rst]: Serial comma nit
```suggestion
``clustermesh-apiserver`` process, the ``kvstoremesh`` process (if enabled),
```

File: Documentation/observability/metrics.rst
Comment (zacharysarah) [inline Documentation/observability/metrics.rst]: Use present tense
```suggestion
passing an empty IP (e.g. ``:9964``) binds the server to all available
```

File: Documentation/observability/metrics.rst
Comment (zacharysarah) [inline Documentation/observability/metrics.rst]: One what? Please verify.
```suggestion
interfaces (there is usually only one interface in a container).
```

**Review by zacharysarah — APPROVED** [review]
Some small changes required, otherwise LGTM. Approving to unblock with the understanding that changes are required prior to merge.

File: kvstoremesh/.gitignore
Comment (YutaroHayakawa) [inline kvstoremesh/.gitignore:1]: Nit: It's a bit odd to me that the name of the executable is `kvstoremesh` maybe `kvstoremesh-apiserver` or `kvstore-agent`?

**Review by YutaroHayakawa — APPROVED** [review]
LGTM 👍 Beautiful code!

File: kvstoremesh/metrics/metrics.go
Comment (giorio94) [reply kvstoremesh/metrics/metrics.go:80]: >  So from what I can tell, with reuising the existing pkg/metrics declarations this will result in using the "cilium_" namespace?

As for the cilium specific metrics, the [namespace is changed to `cilium_kvstoremesh`](https://github.com/cilium/cilium/blob/96abcb95bc34356c9465de536d35749039419815/kvstoremesh/main.go#L45) (and `cilium_clustermesh_apiserver` in the clustermesh-apiserver case). The only metrics that are currently not namespaced in those two cases are the generic ones (go and process).

File: kvstoremesh/.gitignore
Comment (giorio94) [reply kvstoremesh/.gitignore:1]: Honestly I was doubtful about which name to pick, and I ended up using plain *kvstoremesh* mainly for brevity. I'm personally a bit against `kvstoremesh-apiserver` because it is not actually an apiserver (we already have the same problem with the clustermesh-apiserver, which makes understanding its role a bit difficult). The same applies for `kvstoremesh-operator`, because it is not strictly an operator in the kubernetes sense. I don't have strong opinion though.

File: kvstoremesh/metrics/metrics.go
Comment (giorio94) [reply kvstoremesh/metrics/metrics.go:91]: Added, thanks!

**Comment by giorio94** [comment]
Rebased onto main to pick the fixes for conformance-ginkgo

File: pkg/defaults/defaults.go
Comment (tklauser) [inline pkg/defaults/defaults.go]: This value is already used for Hubble Relay's gops port:

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/Documentation/operations/system_requirements.rst?plain=1#L431

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/pkg/hubble/relay/defaults/defaults.go#L20-L21

I guess this should be 9894 instead.

Would also be nice to document the new value in the table:

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/Documentation/operations/system_requirements.rst?plain=1#L413-L436

File: kvstoremesh/option/config.go
Comment (tklauser) [inline kvstoremesh/option/config.go:17]: Would be nice to document this in the following table along with the other pprof ports:

https://github.com/cilium/cilium/blob/68b3ae1918b1bd81e2f82d4fb83c896f8fae4f9f/Documentation/operations/system_requirements.rst?plain=1#L413-L436

File: pkg/defaults/defaults.go
Comment (giorio94) [reply pkg/defaults/defaults.go]: KVStoreMesh is not running in host network, hence it does not risk conflicting with the ports used on the nodes. Do you think it is better to add it to the system requirements also in that case? (I see that clustermesh-apiserver is present there, although it's unclear to me whether that is intentional or due to historical reasons).

Comment (tklauser) [reply pkg/defaults/defaults.go]: I'd still suggest to keep the port values distinct for consistency.

Feel free to omit it from the table, that one admittedly is a bit misleading probably for historical reasons and could use a cleanup.

Comment (giorio94) [reply pkg/defaults/defaults.go]: I've changed the port value to 9894 and opened https://github.com/cilium/cilium/pull/26230 for the cleanup. PTAL

**Review by tklauser — APPROVED** [review]
Thanks!

**Comment by giorio94** [comment]
Rebased onto main to fix CI failure

**Review by michi-covalent — APPROVED** [review]
mostly rubber stamping on behalf of the tophat team. lgtm assuming it's getting reviewed by the actual code owner (sig-clustermesh team) ✅

**Comment by giorio94** [comment]
@tommyp1ckles PTAL

**Review by aditighag — APPROVED** [review]
@giorio94 It's not obvious if there are any policy related changes in the PR. It looks like sig-policy review was requested after this change - https://github.com/cilium/cilium/compare/9a82196551f3ce7b212c8884c3a5b62895d1d489..d5abdcadbb249cc901ae0511ab332d3346864ee8. 
Let me know if that's not the case.

File: kvstoremesh/metrics/metrics.go
Comment (tommyp1ckles) [reply kvstoremesh/metrics/metrics.go:80]: ah I see, missed that line, thought this would be the same as the agent. Ok looks good thanks.

**Review by tommyp1ckles — APPROVED** [review]
Metrics LGTM 🙏

**Comment by giorio94** [comment]
> @giorio94 It's not obvious if there are any policy related changes in the PR. It looks like sig-policy review was requested after this change - https://github.com/cilium/cilium/compare/9a82196551f3ce7b212c8884c3a5b62895d1d489..d5abdcadbb249cc901ae0511ab332d3346864ee8. Let me know if that's not the case.

Yeah, there's no policy-related change. It's only the metrics documentation page which seems to be also owned by sig-policy. Thanks!

**Pull request merged**
